### ISSUE TITLE
Problems loading GO

### ISSUE THREAD

**Issue by Nic-Nic** [saezlab/pypath#79]
Hi Dénes,
I was trying to load both the `intercell` and `annotations` data but both raised an error.

For the intercellular, when running `annot.AnnotationTable(keep_annotators=True)` I get problems with the JSON decoder. Traceback is:
```python
---------------------------------------------------------------------------
JSONDecodeError                           Traceback (most recent call last)
<ipython-input-17-473ebc168951> in <module>
----> 1 a = annot.AnnotationTable(keep_annotators=True)

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in __init__(self, uniprots, use_sources, use_fields, ncbi_tax_id, swissprot_only, keep_annotators, load)
   1242         if load:
   1243 
-> 1244             self.load()
   1245 
   1246 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in load(self)
   1266 
   1267             annot = getattr(self._module, cls)(
-> 1268                 ncbi_tax_id = self.ncbi_tax_id
   1269             )
   1270 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in __init__(self, categories, go_annot, ncbi_tax_id)
    970             self,
    971             name = 'GO_Intercell',
--> 972             ncbi_tax_id = ncbi_tax_id,
    973         )
    974 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in __init__(self, name, ncbi_tax_id, input_method, input_args, **kwargs)
    317         )
    318 
--> 319         self.load()
    320 
    321 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in load(self)
    984             categories = self.categories,
    985             go_annot = self.go_annot,
--> 986             ncbi_tax_id = self.ncbi_tax_id,
    987         )
    988 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in __init__(self, categories, go_annot, ncbi_tax_id)
    946             categories = categories,
    947             go_annot = go_annot,
--> 948             ncbi_tax_id = ncbi_tax_id,
    949         )
    950 

/usr/local/lib/python3.7/dist-packages/pypath/go.py in __init__(self, categories, go_annot, ncbi_tax_id)
    847         session_mod.Logger.__init__(self, name = 'go')
    848 
--> 849         self.go_annot = go_annot or get_db() # TODO: consider ncbi_tax_id at
    850                                              # selection DB
    851 

/usr/local/lib/python3.7/dist-packages/pypath/go.py in get_db()
    999     if 'db' not in globals():
   1000 
-> 1001         init_db()
   1002 
   1003     return globals()['db']

/usr/local/lib/python3.7/dist-packages/pypath/go.py in init_db()
    984     """
    985 
--> 986     globals()['db'] = GOAnnotation()
    987 
    988 

/usr/local/lib/python3.7/dist-packages/pypath/go.py in __init__(self, organism, ontology)
    345         session_mod.Logger.__init__(self, name = 'go')
    346 
--> 347         self.ontology = ontology or GeneOntology()
    348 
    349         annot = dataio.go_annotations_goa(organism = organism)

/usr/local/lib/python3.7/dist-packages/pypath/go.py in __init__(self)
     62         session_mod.Logger.__init__(self, name = 'go')
     63 
---> 64         self._load()
     65 
     66 

/usr/local/lib/python3.7/dist-packages/pypath/go.py in _load(self)
     77     def _load(self):
     78 
---> 79         self._load_terms()
     80         self._load_tree()
     81         self._set_aspect()

/usr/local/lib/python3.7/dist-packages/pypath/go.py in _load_terms(self)
     88     def _load_terms(self):
     89 
---> 90         self._terms = dataio.go_terms_quickgo()
     91 
     92 

/usr/local/lib/python3.7/dist-packages/pypath/dataio.py in go_terms_quickgo(aspects)
   4217         c = curl.Curl(page_url, silent = True)
   4218 
-> 4219         this_result = json.loads(c.result)
   4220         last_page = this_result['pageInfo']['total']
   4221 

/usr/lib/python3.7/json/__init__.py in loads(s, encoding, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)
    346             parse_int is None and parse_float is None and
    347             parse_constant is None and object_pairs_hook is None and not kw):
--> 348         return _default_decoder.decode(s)
    349     if cls is None:
    350         cls = JSONDecoder

/usr/lib/python3.7/json/decoder.py in decode(self, s, _w)
    335 
    336         """
--> 337         obj, end = self.raw_decode(s, idx=_w(s, 0).end())
    338         end = _w(s, end).end()
    339         if end != len(s):

/usr/lib/python3.7/json/decoder.py in raw_decode(self, s, idx)
    351         """
    352         try:
--> 353             obj, end = self.scan_once(s, idx)
    354         except StopIteration as err:
    355             raise JSONDecodeError("Expecting value", s, err.value) from None

JSONDecodeError: Unterminated string starting at: line 1 column 97945 (char 97944)
```

When loading the annotations with `intercell.IntercellAnnotation()`:

```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-27-1b0e9044a0e5> in <module>
----> 1 i = intercell.IntercellAnnotation()

/usr/local/lib/python3.7/dist-packages/pypath/intercell.py in __init__(self, class_definitions)
     48         annot.CustomAnnotation.__init__(
     49             self,
---> 50             class_definitions = class_definitions,
     51         )
     52 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in __init__(self, class_definitions)
     87         session_mod.Logger.__init__(self, name = 'annot')
     88 
---> 89         self.annotdb = get_db()
     90 
     91         self._class_definitions = {}

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in get_db(keep_annotators)
   1474     if 'db' not in globals():
   1475 
-> 1476         init_db(keep_annotators = keep_annotators)
   1477 
   1478     return globals()['db']

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in init_db(keep_annotators)
   1463     """
   1464 
-> 1465     globals()['db'] = AnnotationTable(keep_annotators = keep_annotators)
   1466 
   1467 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in __init__(self, uniprots, use_sources, use_fields, ncbi_tax_id, swissprot_only, keep_annotators, load)
   1242         if load:
   1243 
-> 1244             self.load()
   1245 
   1246 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in load(self)
   1266 
   1267             annot = getattr(self._module, cls)(
-> 1268                 ncbi_tax_id = self.ncbi_tax_id
   1269             )
   1270 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in __init__(self, categories, go_annot, ncbi_tax_id)
    970             self,
    971             name = 'GO_Intercell',
--> 972             ncbi_tax_id = ncbi_tax_id,
    973         )
    974 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in __init__(self, name, ncbi_tax_id, input_method, input_args, **kwargs)
    317         )
    318 
--> 319         self.load()
    320 
    321 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in load(self)
    987         )
    988 
--> 989         annot_uniprots = annot.get_annotations()
    990 
    991         _annot = collections.defaultdict(set)

/usr/local/lib/python3.7/dist-packages/pypath/go.py in get_annotations(self, uniprots)
    937                 self.get_annotation(category, uniprots = uniprots)
    938             )
--> 939             for category in self.categories.keys()
    940         )
    941 

/usr/local/lib/python3.7/dist-packages/pypath/go.py in <genexpr>(.0)
    937                 self.get_annotation(category, uniprots = uniprots)
    938             )
--> 939             for category in self.categories.keys()
    940         )
    941 

/usr/local/lib/python3.7/dist-packages/pypath/go.py in get_annotation(self, category, uniprots)
    919             self.categories[category],
    920             uniprots = uniprots,
--> 921             return_uniprots = True,
    922         )
    923 

/usr/local/lib/python3.7/dist-packages/pypath/go.py in select(self, terms, uniprots, return_uniprots)
    781         if isinstance(terms, common.basestring) and len(terms) > 10:
    782 
--> 783             result = self.select_by_expr(terms, uniprots = uniprots)
    784 
    785         # either one term or a set of terms

/usr/local/lib/python3.7/dist-packages/pypath/go.py in select_by_expr(self, expr, uniprots, return_uniprots)
    630             expr = expr,
    631             uniprots = uniprots,
--> 632             return_uniprots = return_uniprots,
    633         )
    634 

/usr/local/lib/python3.7/dist-packages/pypath/go.py in select_by_expr_terms(self, expr, uniprots, return_uniprots)
    705 
    706             # we do actual processing of the expression
--> 707             elif it.lower() == 'not':
    708 
    709                 # token is negation

AttributeError: 'NoneType' object has no attribute 'lower'
```

Do you also get this or am I missing something?

Thanks for your help

**Comment by deeenes** [comment]
Hi Nico,

The first looks like an interrupted download. Try in context `with pypath.curl.cache_off()`. For the second I added an error handling in cb2e5c4 so you will see what's going on and also to avoid raising exception on any failed processing.

**Comment by Nic-Nic** [comment]
Thanks, will try :)

**Comment by Nic-Nic** [comment]
Annotation now gives this (using `cache_off()`):

```python
<class 'json.decoder.JSONDecodeError'>, Expecting ',' delimiter: line 1 column 293179 (char 293178), <traceback object at 0x7f13d29a8548>
---------------------------------------------------------------------------
JSONDecodeError                           Traceback (most recent call last)
<ipython-input-12-686be4798d49> in <module>
      1 with pypath.curl.cache_off():
----> 2     a = annot.AnnotationTable(keep_annotators=True)
      3     a.load()

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in __init__(self, uniprots, use_sources, use_fields, ncbi_tax_id, swissprot_only, complexes, keep_annotators, load)
   1343         if load:
   1344 
-> 1345             self.load()
   1346 
   1347 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in load(self)
   1367 
   1368             annot = getattr(self._module, cls)(
-> 1369                 ncbi_tax_id = self.ncbi_tax_id
   1370             )
   1371 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in __init__(self, categories, go_annot, ncbi_tax_id)
    977             self,
    978             name = 'GO_Intercell',
--> 979             ncbi_tax_id = ncbi_tax_id,
    980         )
    981 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in __init__(self, name, ncbi_tax_id, input_method, input_args, **kwargs)
    324         )
    325 
--> 326         self.load()
    327 
    328 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in load(self)
    991             categories = self.categories,
    992             go_annot = self.go_annot,
--> 993             ncbi_tax_id = self.ncbi_tax_id,
    994         )
    995 

/usr/local/lib/python3.7/dist-packages/pypath/annot.py in __init__(self, categories, go_annot, ncbi_tax_id)
    953             categories = categories,
    954             go_annot = go_annot,
--> 955             ncbi_tax_id = ncbi_tax_id,
    956         )
    957 

/usr/local/lib/python3.7/dist-packages/pypath/go.py in __init__(self, categories, go_annot, ncbi_tax_id)
    860         session_mod.Logger.__init__(self, name = 'go')
    861 
--> 862         self.go_annot = go_annot or get_db() # TODO: consider ncbi_tax_id at
    863                                              # selection DB
    864 

/usr/local/lib/python3.7/dist-packages/pypath/go.py in get_db()
   1012     if 'db' not in globals():
   1013 
-> 1014         init_db()
   1015 
   1016     return globals()['db']

/usr/local/lib/python3.7/dist-packages/pypath/go.py in init_db()
    997     """
    998 
--> 999     globals()['db'] = GOAnnotation()
   1000 
   1001 

/usr/local/lib/python3.7/dist-packages/pypath/go.py in __init__(self, organism, ontology)
    345         session_mod.Logger.__init__(self, name = 'go')
    346 
--> 347         self.ontology = ontology or GeneOntology()
    348 
    349         annot = dataio.go_annotations_goa(organism = organism)

/usr/local/lib/python3.7/dist-packages/pypath/go.py in __init__(self)
     62         session_mod.Logger.__init__(self, name = 'go')
     63 
---> 64         self._load()
     65 
     66 

/usr/local/lib/python3.7/dist-packages/pypath/go.py in _load(self)
     78 
     79         self._load_terms()
---> 80         self._load_tree()
     81         self._set_aspect()
     82         self._set_name()

/usr/local/lib/python3.7/dist-packages/pypath/go.py in _load_tree(self)
     97         )
     98         self.descendants = self._merge_aspects(
---> 99             dataio.go_descendants_quickgo()
    100         )
    101 

/usr/local/lib/python3.7/dist-packages/pypath/dataio.py in go_descendants_quickgo(aspects, terms, relations)
   4108             )
   4109 
-> 4110             result = json.load(c.fileobj)
   4111 
   4112             for res in result['results']:

/usr/lib/python3.7/json/__init__.py in load(fp, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)
    294         cls=cls, object_hook=object_hook,
    295         parse_float=parse_float, parse_int=parse_int,
--> 296         parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
    297 
    298 

/usr/lib/python3.7/json/__init__.py in loads(s, encoding, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)
    346             parse_int is None and parse_float is None and
    347             parse_constant is None and object_pairs_hook is None and not kw):
--> 348         return _default_decoder.decode(s)
    349     if cls is None:
    350         cls = JSONDecoder

/usr/lib/python3.7/json/decoder.py in decode(self, s, _w)
    335 
    336         """
--> 337         obj, end = self.raw_decode(s, idx=_w(s, 0).end())
    338         end = _w(s, end).end()
    339         if end != len(s):

/usr/lib/python3.7/json/decoder.py in raw_decode(self, s, idx)
    351         """
    352         try:
--> 353             obj, end = self.scan_once(s, idx)
    354         except StopIteration as err:
    355             raise JSONDecodeError("Expecting value", s, err.value) from None

JSONDecodeError: Expecting ',' delimiter: line 1 column 293179 (char 293178)
```
And intercellular gives a similar one (w/o disabling the cache) and if I disable it then gives:

```python
<class 'AttributeError'>, 'NoneType' object has no attribute 'split', <traceback object at 0x7f5f90606d88>
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-28-468069dc5e4e> in <module>
      1 with pypath.curl.cache_off():
----> 2     i = intercell.IntercellAnnotation()

~/.local/lib/python3.7/site-packages/pypath/intercell.py in __init__(self, class_definitions)
     48         annot.CustomAnnotation.__init__(
     49             self,
---> 50             class_definitions = class_definitions,
     51         )
     52 

~/.local/lib/python3.7/site-packages/pypath/annot.py in __init__(self, class_definitions)
     94         session_mod.Logger.__init__(self, name = 'annot')
     95 
---> 96         self.annotdb = get_db()
     97 
     98         self._class_definitions = {}

~/.local/lib/python3.7/site-packages/pypath/annot.py in get_db(keep_annotators)
   1575     if 'db' not in globals():
   1576 
-> 1577         init_db(keep_annotators = keep_annotators)
   1578 
   1579     return globals()['db']

~/.local/lib/python3.7/site-packages/pypath/annot.py in init_db(keep_annotators)
   1564     """
   1565 
-> 1566     globals()['db'] = AnnotationTable(keep_annotators = keep_annotators)
   1567 
   1568 

~/.local/lib/python3.7/site-packages/pypath/annot.py in __init__(self, uniprots, use_sources, use_fields, ncbi_tax_id, swissprot_only, complexes, keep_annotators, load)
   1332                 dataio.all_uniprots(
   1333                     organism = ncbi_tax_id,
-> 1334                     swissprot = swissprot_only,
   1335                 )
   1336             )

~/.local/lib/python3.7/site-packages/pypath/dataio.py in all_uniprots(organism, swissprot)
    207 def all_uniprots(organism = 9606, swissprot = None):
    208 
--> 209     return uniprot_input.all_uniprots(organism, swissprot)
    210 
    211 

~/.local/lib/python3.7/site-packages/pypath/uniprot_input.py in all_uniprots(organism, swissprot)
     41 
     42     return [
---> 43         l.strip() for l in data.split('\n')[1:] if l.strip()
     44     ]

AttributeError: 'NoneType' object has no attribute 'split'
```

:disappointed:

**Comment by deeenes** [comment]
GO changed some terms recently, this fixed in dbac750 and for me it loads without error. The JSON error is still due to download error, e.g. HTTP timeout from either side, in 8439de2 I try to give more chance for it to work with slow networks. The `uniprot_input.all_uniprots()` also might be due to timeout, but for me this also works, and it consists of only one query: `https://www.uniprot.org/uniprot/?query=organism%3A9606&format=tab&columns=id`, I have no idea how to split this into smaller queries.

**Comment by Nic-Nic** [comment]
Solved
